xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_backward.h

```diff
@@ -1059,6 +1082,97 @@ struct AttentionBackwardKernel {
         }));
   }
 
+  template <int kElementsPerAccess>
+  static __device__ void computeDelta(Params const& p, int32_t query_start) {
+    // Each thread computes one value for Delta
+    // Depending on warp configuration, we might have multiple
+    // threads of the same warp working on the same row
+    using AccessType = cutlass::Array<scalar_t, kElementsPerAccess>;
+    static_assert(kNumThreads >= kBlockSizeI);
+    static constexpr int kNumThreadsPerLine = kNumThreads / kBlockSizeI;
+    int16_t thread_id = get_thread_id();
+
+    int16_t laneFirstCol =
+        kElementsPerAccess * (get_lane_id() % kNumThreadsPerLine);
+    int16_t laneRow = thread_id / kNumThreadsPerLine;
+
+    const __restrict__ AccessType* grad_output_ptr =
+        reinterpret_cast<const __restrict__ AccessType*>(
+            p.grad_output_ptr + (query_start + laneRow) * p.head_dim_value +
+            laneFirstCol);
+    const __restrict__ AccessType* output_ptr =
+        reinterpret_cast<const __restrict__ AccessType*>(
+            p.output_ptr + (query_start + laneRow) * p.head_dim_value +
+            laneFirstCol);
+
+    static constexpr int64_t kMaxIters =
+        kMaxK / (kElementsPerAccess * kNumThreadsPerLine);
+    constexpr int kPipelineStages = 2;
+    accum_t delta_value = accum_t(0);
+    using GlobalLoad =
+        cutlass::arch::global_load<AccessType, sizeof(AccessType)>;
+    AccessType frag_grad_output[kPipelineStages];
+    AccessType frag_output[kPipelineStages];
+
+    auto loadAndIncrement = [&](int ld_pos, bool pred) {
+      frag_grad_output[ld_pos].clear();
+      frag_output[ld_pos].clear();
+      GlobalLoad(frag_grad_output[ld_pos], grad_output_ptr, pred);
+      GlobalLoad(frag_output[ld_pos], output_ptr, pred);
+      grad_output_ptr += kNumThreadsPerLine;
+      output_ptr += kNumThreadsPerLine;
+    };
+
+    CUTLASS_PRAGMA_UNROLL
+    for (int iter = 0; iter < kPipelineStages - 1; ++iter) {
+      int ld_pos = iter % kPipelineStages;
+      bool pred =
+          (laneFirstCol + iter * kElementsPerAccess * kNumThreadsPerLine) <
+          p.head_dim_value;
+      loadAndIncrement(ld_pos, pred);
+    }
+    auto columnIteration = [&](int iter) {
+      // Load for next iter
+      int ld_pos = (iter + kPipelineStages - 1) % kPipelineStages;
+      bool pred = (laneFirstCol +
+                   (iter + kPipelineStages - 1) * kElementsPerAccess *
+                       kNumThreadsPerLine) < p.head_dim_value;
+      loadAndIncrement(ld_pos, pred);
+      CUTLASS_PRAGMA_UNROLL
+      for (int i = 0; i < AccessType::kElements; ++i) {
+        delta_value += accum_t(frag_output[iter % kPipelineStages][i]) *
+            accum_t(frag_grad_output[iter % kPipelineStages][i]);
+      }
+    };
+
+    // If we have a small lower-bound for K, we can unroll the loop
+    if (kMaxK <= 256) {
+      CUTLASS_PRAGMA_UNROLL
+      for (int iter = 0; iter < kMaxIters; ++iter) {
+        columnIteration(iter);
+      }
```

Comment:
Can we ensure that our tests cover well those cases? Including even / odd-sized K ?